test/include/seqan3/test/sandboxed_path.hpp

```diff
@@ -0,0 +1,461 @@
+            }
+            current_pos = end_pos;
+        }
+        std::string normalised_path;
+        for (auto const& p : path_parts) {
+            normalised_path += "/" + p;
+        }
+        std::filesystem::path::operator=(normalised_path.data());
+#else
+        auto normalised_path = std::filesystem::weakly_canonical(sandbox_path / *this);
+        std::filesystem::path::operator=(normalised_path);
+#endif
+    }
+
+public:
+    /*!\brief Checks the invariant.
+     *
+     * Checks that the invariant of the class sandboxed_path is kept.
+     * See class description for invariant.
+     * \throws std::filesystem::filesystem_error if invariant was violated.
+     */
+    void checkInvariant() const
+    {
+        // Checking that sandbox_path is an absolute path
+        if (!sandbox_path.is_absolute()) {
+            throw std::filesystem::filesystem_error("sandbox path must be an absolute path",
+                                                sandbox_path, *this,
+                                                std::make_error_code(std::errc::invalid_argument));
+        }
+
+#if SEQAN3_WORKAROUND_GCC7_INCOMPLETE_FILESYSTEM
+        auto current_dir = string();
+        auto sandbox_dir = sandbox_path.string();
+
+        // remove trailing '/' of path
+        if (!sandbox_dir.empty() and sandbox_dir.back() == '/') {
+            sandbox_dir.pop_back();
+        }
+
+        // Leaving the temporary directory is not allowed.
+        bool starts_with_sandbox_dir = (current_dir.rfind(sandbox_dir, 0) == 0);
+        if (!starts_with_sandbox_dir or
+            (current_dir.size() > sandbox_dir.size() and current_dir.at(sandbox_dir.size()) != '/'))
+        {
+            throw std::filesystem::filesystem_error("Leaving temporary directory is not allowed!",
+                                                sandbox_path, *this,
+                                                std::make_error_code(std::errc::invalid_argument));
+        }
+
+#else
+        auto rel_path = lexically_relative(sandbox_path);
+
+        // Leaving the temporary directory is not allowed.
+        if (rel_path.string().find("..") == 0) {
+            throw std::filesystem::filesystem_error("Leaving temporary directory is not allowed!",
+                                                sandbox_path, *this,
+                                                std::make_error_code(std::errc::invalid_argument));
+        }
+#endif
+    }
+
+    /*!\brief Replaces the path with a new path.
+     * \tparam source_t The type of the source.
+     *
+     * This works the same as std::filesystem::path::assign
+     * and additionally checks the invariant.
+     *
+     * ### Exceptions
+     *
+     * Basic exception guarantee
+     */
+    template <typename source_t>
+    sandboxed_path & assign(source_t const & source)
+    {
+        std::filesystem::path::assign(source);
+        normalise();
+        checkInvariant();
+        return *this;
+    }
```

Comment:
Maybe you can think of another name than `source` and `source_t`?
It's from the standard, but it would be probably more clear if we used `path_t` and `new_path` or something like this?